compiler/rustc_session/src/config/cfg.rs

```diff
@@ -379,3 +380,54 @@ impl CheckCfg {
         ins!(sym::windows, no_values);
     }
 }
+
+pub(crate) fn disallow_cfgs(sess: &Session, user_cfgs: &Cfg) {
+    let disallow = |cfg: &(Symbol, Option<Symbol>), controlled_by| {
+        let cfg_name = cfg.0;
+        let cfg = if let Some(value) = cfg.1 {
+            format!(r#"{}="{}""#, cfg_name, value)
+        } else {
+            format!("{}", cfg_name)
+        };
+        sess.dcx().emit_fatal(DisallowConfig { cfg, cfg_name, controlled_by })
+    };
+
+    // We want to restrict setting cfgs that will produce "incoherent" behavior between
+    // the cfg and the "real" flag that sets it, but not all cfgs produce incoherent
+    // behavior, we therefore exclude those cfgs:
+    //
+    //  - test
+    //  - clippy
+    //  - doc
+    //  - doctest
+    //  - miri
+    //  - rustfmt
+    //  - overflow_checks
+    //  - debug_assertions
+    //  - ub_checks
```

Comment:
All of these may produce incoherent behavior, from what I see.
E.g. `test` and `doctest` are used in standard library to disable some panic machinery, if they are passed incorrectly then things will break.
So I don't see any fundamental differences from the configs above and below.
(Except maybe rustfmt or clippy, because I'm not sure why they exist at all.)

Perhaps they should all produce lint warnings.
The warnings can be silenced with `-A`/`#[allow]` if the user really needs it.

Reply:
Of course, I've changed the implementation to a deny-by-default lint and included all possible cfgs (I've left some links in the code to explain why some are excluded).

(I had to modify the `BufferedEarlyLint` struct to contain an optional span, I can submit this patch as a separate PR if desired)